overlord/snapstate/snapstate.go

```diff
@@ -1917,19 +1917,21 @@ func AutoRefresh(ctx context.Context, st *state.State) ([]string, []*state.TaskS
 	}
 
 	// TODO: rename to autoRefreshTasks when old auto refresh logic gets removed.
-	return autoRefreshPhase1(ctx, st)
+	return autoRefreshPhase1(ctx, st, "")
 }
 
 // autoRefreshPhase1 creates gate-auto-refresh hooks and conditional-auto-refresh
-// task that initiates actual refresh.
+// task that initiates actual refresh. forGatingSnap is optional and limits auto-refresh
+// to the snaps affecting the given snap only; it defaults to all snaps if nil.
 // The state needs to be locked by the caller.
-func autoRefreshPhase1(ctx context.Context, st *state.State) ([]string, []*state.TaskSet, error) {
+func autoRefreshPhase1(ctx context.Context, st *state.State, forGatingSnap string) ([]string, []*state.TaskSet, error) {
 	user, err := userFromUserID(st, 0)
 	if err != nil {
 		return nil, nil, err
 	}
 
 	refreshOpts := &store.RefreshOptions{IsAutoRefresh: true}
+	// XXX: should we skip refreshCandidates if forGatingSnap isn't empty (meaning we're handling proceed from a snap)?
```

Comment:
yes, this is not great because it means we trigger a store op

Reply:
Hmm, so avoiding this turns out to require some potentially annoying changes, e.g. affectedByRefresh() needs snap.Infos (that comes from candidates/updates) as it looks into info.Hooks. It should be possible but maybe as a followup?

Reply:
maybe it doesn't matter because we will do store operations anyway, also we do need to refresh assertions to do the right thing with validations. It's another reason pointing to the fact that maybe calling proceed outside of the normal refresh hook should be a privileged operation.